test-app/tests/helpers/file-queue-helper-test.js

```diff
@@ -110,6 +113,90 @@ module('Integration | Helper | file-queue', function (hooks) {
     assert.dom('[data-test-file]').doesNotExist();
   });
 
+  test('will be notified when an upload starts', async function (assert) {
+    this.server.post(
+      '/upload-file',
+      uploadHandler(() => {})
+    );
+
+    this.upload = (file) => file.upload('/upload-file');
+
+    this.uploadStarted = (file) =>
+      assert.step(`upload started: ${file.name}, state: ${file.state}`);
+
+    await render(hbs`
+      {{#let (file-queue onFileAdded=this.upload onUploadStarted=this.uploadStarted) as |queue|}}
+        <label>
+          <input type="file" {{queue.selectFile}}>
+          Select File
+        </label>
+      {{/let}}
+    `);
+
+    await selectFiles('input[type=file]', new File([], 'dingus.txt'));
+
+    assert.verifySteps(['upload started: dingus.txt, state: uploading']);
+  });
+
+  test('will be notified when an upload is successful', async function (assert) {
+    this.server.post(
+      '/upload-file',
+      uploadHandler(() => {})
+    );
+
+    this.upload = (file) => file.upload('/upload-file');
+
+    this.uploadSucceeded = (file, response) =>
+      assert.step(
+        `upload succeeded: ${file.name}, state: ${file.state}, response.status: ${response.status}`
+      );
+
+    await render(hbs`
+      {{#let (file-queue onFileAdded=this.upload onUploadSucceeded=this.uploadSucceeded) as |queue|}}
+        <label>
+          <input type="file" {{queue.selectFile}}>
+          Select File
+        </label>
+      {{/let}}
+    `);
+
+    await selectFiles('input[type=file]', new File([], 'dingus.txt'));
+
+    assert.verifySteps([
+      'upload succeeded: dingus.txt, state: uploading, response.status: 201',
+    ]);
```

Comment:
As explained in (1) the state of this file is `FileState.Uploading` which is kind of unexpected due to this being the `onUploadSucceeded` callback.

Reply:
I would tend towards fixing that if possible. It may be very confusing to the user. And it makes some use cases more challenging. E.g. a user can not handle both settled states (`onUploadSucceeded`, `onUploadFailed`) with the same handler if needing to know at any time processing the files of was successful or not.

Reply:
> when state is updated they'll be removed from the queue 

Is it a problem if the file is removed from queue when `onUploadSucceeded` and `onUploadFailed` is called?

Reply:
> I would tend towards fixing that if possible. It may be very confusing to the user

Agreed.

> Is it a problem if the file is removed from queue when onUploadSucceeded and onUploadFailed is called?

No I don't think that would cause any problems.